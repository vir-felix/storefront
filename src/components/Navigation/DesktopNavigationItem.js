import React, { Component, Fragment } from "react";
import PropTypes from "prop-types";
import Button from "material-ui/Button";
import Divider from "material-ui/Divider";
import Grid from "material-ui/Grid";
import ListItemText from "material-ui/List/ListItemText";
import MenuList from "material-ui/Menu/MenuList";
import MenuItem from "material-ui/Menu/MenuItem";
import Popover from "material-ui/Popover";
import ChevronDownIcon from "mdi-material-ui/ChevronDown";
import ChevronUpIcon from "mdi-material-ui/ChevronUp";

import { observer } from "mobx-react";
import { action, computed, observable } from "mobx";
import { withStyles } from "material-ui/styles";
import { Router } from "../../routes";

const styles = (theme) => ({
  popover: {
    maxWidth: "100vw",
    padding: theme.spacing.unit * 2,
    width: "100vw"
  },
  grid: {
    width: "100vw"
  }
});

@withStyles(styles)
@observer
class DesktopNavigationItem extends Component {
  static propTypes = {
    classes: PropTypes.object,
    navItem: PropTypes.object
  };

  static defaultProps = {
    classes: {},
    navItem: {}
  };

  get hasSubNavItems() {
    const { navItem: { relatedTags } } = this.props;
    return Array.isArray(relatedTags) && relatedTags.length > 0;
  }

  @observable _isSubNavOpen = false;

  @computed
  get isSubNavOpen() {
    return this._isSubNavOpen;
  }

  set isSubNavOpen(value) {
    this._isSubNavOpen = value;
  }

  @action
  onClick = () => {
    const { navItem } = this.props;

    if (this.hasSubNavItems) {
      this.isSubNavOpen = !this.isSubNavOpen;
    } else {
      Router.pushRoute("tag", { slug: navItem.name });
    }
  };

  @action
  onClose = () => {
    this.isSubNavOpen = false;
  };

  renderSubNav(navItemGroup) {
    return (
      <Fragment>
        <Divider />
        {navItemGroup.relatedTags.map((navItem, index) => (
          <MenuItem dense key={index}>
            <ListItemText primary={navItem.title} />
          </MenuItem>
        ))}
      </Fragment>
    );
  }

  renderPopover() {
    const { classes, navItem: { relatedTags } } = this.props;
    return (
      <Popover
        classes={{ paper: classes.popover }}
        anchorReference={"anchorPosition"}
        anchorPosition={{ top: 64 }}
        elevation={1}
        onClose={this.onClose}
        open={this.isSubNavOpen}
      >
        <Grid container className={classes.grid} spacing={16}>
          {relatedTags.map((navItemGroup, index) => (
            <Grid item key={index}>
              <MenuList disablePadding>
                <MenuItem>
                  <ListItemText primary={navItemGroup.title} />
                </MenuItem>
                {Array.isArray(navItemGroup.relatedTags) && this.renderSubNav(navItemGroup)}
              </MenuList>
            </Grid>
          ))}
        </Grid>
      </Popover>
    );
  }

  render() {
    const { navItem } = this.props;
    return (
      <Fragment>
        <Button color="inherit" onClick={this.onClick}>
          {navItem.name}
          {this.hasSubNavItems && <Fragment>{this.isSubNavOpen ? <ChevronUpIcon /> : <ChevronDownIcon />}</Fragment>}
        </Button>
        {this.hasSubNavItems && this.renderPopover()}
      </Fragment>
    );
  }
}

export default DesktopNavigationItem;
