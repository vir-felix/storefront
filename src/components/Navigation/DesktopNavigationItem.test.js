import React from "react";
import renderer from "react-test-renderer";
import DesktopNavigationItem from "./DesktopNavigationItem";

const testMenuItem = {
  _id: "123",
  name: "Shop",
  slug: "shop"
};

const testMenuItemWithRelatedTags = {
  _id: "123",
  name: "Shop",
  slug: "shop",
  relatedTags: [
    { _id: "111", name: "Clothes", slug: "clothes" },
    { _id: "222", name: "Shoes", slug: "shoes" },
    { _id: "333", name: "Accessories", slug: "accessories" }
  ]
};

test("basic snapshot", () => {
  const component = renderer.create(<DesktopNavigationItem navItems={testMenuItem} />);
  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

test("snapshot with related items", () => {
  const component = renderer.create(<DesktopNavigationItem navItems={testMenuItemWithRelatedTags} />);
  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});
