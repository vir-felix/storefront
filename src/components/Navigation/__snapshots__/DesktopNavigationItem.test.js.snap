// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`basic snapshot 1`] = `
<button
  className="MuiButtonBase-root-18 MuiButton-root-3 MuiButton-colorInherit-7"
  disabled={false}
  onBlur={[Function]}
  onClick={[Function]}
  onFocus={[Function]}
  onKeyDown={[Function]}
  onKeyUp={[Function]}
  onMouseDown={[Function]}
  onMouseLeave={[Function]}
  onMouseUp={[Function]}
  onTouchEnd={[Function]}
  onTouchMove={[Function]}
  onTouchStart={[Function]}
  tabIndex="0"
  type="button"
>
  <span
    className="MuiButton-label-4"
  />
  <span
    className="MuiTouchRipple-root-21"
  />
</button>
`;

exports[`snapshot with related items 1`] = `
<button
  className="MuiButtonBase-root-18 MuiButton-root-3 MuiButton-colorInherit-7"
  disabled={false}
  onBlur={[Function]}
  onClick={[Function]}
  onFocus={[Function]}
  onKeyDown={[Function]}
  onKeyUp={[Function]}
  onMouseDown={[Function]}
  onMouseLeave={[Function]}
  onMouseUp={[Function]}
  onTouchEnd={[Function]}
  onTouchMove={[Function]}
  onTouchStart={[Function]}
  tabIndex="0"
  type="button"
>
  <span
    className="MuiButton-label-4"
  />
  <span
    className="MuiTouchRipple-root-21"
  />
</button>
`;
